proposals/NNNN-isolated-synchronous-deinit.md

```diff
@@ -0,0 +1,386 @@
+* Implementation: [apple/swift#60057](https://github.com/apple/swift/pull/60057)
+
+## Introduction
+
+This feature lifts restrictions on `deinit` of actors and GAITs imposed by [SE-0327](https://github.com/apple/swift-evolution/blob/main/proposals/0327-actor-initializers.md) by providing runtime support for hopping onto executors in `__deallocating_deinit()`'s.
+
+## Motivation
+
+Combination of automatics reference counting and deterministic deinitialization makes `deinit` in Swift a powerful tool for resource management. It greatly reduces need for `close()`-like methods (`unsubscribe()`, `cancel()`, `shutdown()`, etc.) in the public API. Such methods not only clutter public API, but also introduce a state where object is already unusable but is still referencable.
+
+Restrictions imposed by [SE-0327](https://github.com/apple/swift-evolution/blob/main/proposals/0327-actor-initializers.md) reduce usefulness of explicit `deinit`s in actors and GAITs. Workarounds for these limitations may involve creation of `close()`-like methods, or even manual reference counting, if API should be able to serve several clients.
+
+In cases when `deinit` belongs to a subclass of `UIView` or `UIViewController` which are known to call deinitializer on the main thread, developers may be tempted to silence the diagnostic by adopting `@unchecked Sendable` in types that are not actually  sendable. This undermines concurrency checking by the compiler, and may lead to data races when using incorrectly marked types in other places.
+
+## Proposed solution
+
+... is to allow execution of `deinit` and object deallocation to be the scheduled on the executor, if needed.
+
+Let's consider [examples from SE-0327](https://github.com/apple/swift-evolution/blob/main/proposals/0327-actor-initializers.md#data-races-in-deinitializers):
+
+In case of several instances with shared data isolated on common actor problem is completely eliminated:
+
+```swift
+class NonSendableAhmed { 
+  var state: Int = 0
+}
+
+@MainActor
+class Maria {
+  let friend: NonSendableAhmed
+
+  init() {
+    self.friend = NonSendableAhmed()
+  }
+
+  init(sharingFriendOf otherMaria: Maria) {
+    // While the friend is non-Sendable, this initializer and
+    // and the otherMaria are isolated to the MainActor. That is,
+    // they share the same executor. So, it's OK for the non-Sendable value
+    // to cross between otherMaria and self.
+    self.friend = otherMaria.friend
+  }
+
+  deinit {
+    // Used to be a potential data race.
+    // Now deinit is also isolated on the MainActor.
+    // So this code is perfectly correct.
+    friend.state += 1
+  }
+}
+
+func example() async {
+  let m1 = await Maria()
+  let m2 = await Maria(sharingFriendOf: m1)
+  doSomething(m1, m2)
+} 
+```
+
+In case of escaping self, race condition is eliminated but problem of escaping `self` remains. This problem exists for synchronous code as well and is orthogonal to the concurrency features.
```

Comment:
Actually, it should be "In the case of X, we solve Y, but the problem of Z remains." The fact `X` and `Z` ended up looking the same is an unfortunate coincidence. I think I can re-phrase to avoid it.